File: backend/app/models/certificate.py
import secrets
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


def generate_certificate_number() -> str:
    return f"CERT-{secrets.token_hex(6).upper()}"


class Certificate(Base):
    __tablename__ = "certificates"
    __table_args__ = (UniqueConstraint("user_id", "course_id", name="uq_user_course_certificate"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True, nullable=False)
    course_id: Mapped[int] = mapped_column(ForeignKey("courses.id"), index=True, nullable=False)
    certificate_number: Mapped[str] = mapped_column(String(40), unique=True, index=True, nullable=False)
    issued_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), nullable=False)

    user: Mapped["User"] = relationship(back_populates="certificates")
    course: Mapped["Course"] = relationship(back_populates="certificates")

File: mobile/lib/models/certificate.dart
class Certificate {
  final int id;
  final int courseId;
  final String courseTitle;
  final String certificateNumber;
  final DateTime issuedAt;

  Certificate({
    required this.id,
    required this.courseId,
    required this.courseTitle,
    required this.certificateNumber,
    required this.issuedAt,
  });

  factory Certificate.fromJson(Map<String, dynamic> json) => Certificate(
    id: json['id'] as int,
    courseId: json['course_id'] as int,
    courseTitle: json['course_title'] as String,
    certificateNumber: json['certificate_number'] as String,
    issuedAt: DateTime.parse(json['issued_at'] as String),
  );
}
